lib/Sema/CSDiagnostics.cpp

```diff
@@ -8187,6 +8187,50 @@ bool UnableToInferClosureReturnType::diagnoseAsError() {
   return true;
 }
 
+bool UnableToInferGenericPackElementType::diagnoseAsError() {
+
+  const auto* locator = getLocator();
+  auto path = locator->getPath();
+
+  if (path.size() > 1) {
```

Comment:
The way everything is setup today I don't think you need any of this checking in the `path`. The fix is always going to be with locator that ends on `PackElement` and `simplifyLocator` changes take care of resolving `anchor` correctly, so I think this whole method could be simplified down to:

```c++
auto *locator = getLocator();
auto *elementLoc = locator->castLastElementTo<LocatorPathElt::PackElement>();
if (isExpr<NilLiteralExpr>(getAnchor())) {
  emitDiagnostic(diag::unresolved_nil_literal);
} else {
  // unable to infer the type of an element of generic pack params
  emitDiagnostic(diag::could_not_infer_pack_element, elementLoc->getIndex());
}

if (auto *applyExpr = getAsExpr<ApplyExpr>(locator->getAnchor())) {
   // use getCalleeLocator + getOverloadChoice to find overload and note
}

return true;
```

Reply:
```swift
func patternInstantiationTupleTest1<each T>() -> (repeat Array<each T>) {}

func patternInstantiationConcreteInvalid() {
  let _: Set<Int> = patternInstantiationTupleTest1()
  // expected-error@-1 {{cannot convert value of type '(repeat Array<_>)' to specified type 'Set<Int>'}}

  let _: (Array<Int>, Set<String>) = patternInstantiationTupleTest1() // expected-error {{'(repeat Array<Int, _>)' is not convertible to '(Array<Int>, Set<String>)', tuples have a different number of elements}}
}
```

Seems like this case in `swift/test/Constraints/pack_expansion_types.swift` unintentionally hits `diag::could_not_infer_pack_element` branch without the guarding conditions.

```
Fixes:
  [fix: ignore specified contextual type] @ locator@0x1339e7800 [Call@/Users/lizhen/Projects/swift-debug/boldly.swift:4:21 → contextual type]
  [fix: specify pack element type] @ locator@0x1339e7a68 [Call@/Users/lizhen/Projects/swift-debug/boldly.swift:4:21 → contextual type → tuple type '(/* shape: $T1 */ repeat Array<$T1>)' → tuple type '(_: Set<Int>)' → tuple element #0 → pack element #0]
(failed constraint (/* shape: Pack{<<unresolvedtype>>} */ repeat Array<Pack{<<unresolvedtype>>}>) conv Set<Int> @ locator@0x134015400 [])
(increasing 'user conversion' score by 1 @ locator@0x134015400 [])
(failed constraint (/* shape: Pack{<<unresolvedtype>>} */ repeat Array<Pack{<<unresolvedtype>>}>) bridging conv Set<Int> @ locator@0x134015400 [])
/Users/lizhen/Projects/swift-debug/boldly.swift:4:21: error: cannot convert value of type '(repeat Array<_>)' to specified type 'Set<Int>'
  let _: Set<Int> = patternInstantiationTupleTest1()
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lizhen/Projects/swift-debug/boldly.swift:4:21: error: could not infer pack element #0 from context
  let _: Set<Int> = patternInstantiationTupleTest1()
                    ^
```

Guess I should add some guards like this?

```swift
  const auto applyArgToParamElt = (path.end() - 2)->getAs<LocatorPathElt::ApplyArgToParam>();
  const auto packElementElt = (path.end() - 1)->getAs<LocatorPathElt::PackElement>();
  if (!applyArgToParamElt || !packElementElt) {
    return false;
  }
```

Reply:
I think that's okay, the conversion fails and the element cannot be inferred because of that. For that particular case we could detect that it's mismatch between collections and propagate the element type down to avoid the second diagnostic.